File: sharedse/src/main/java/com/ustadmobile/port/sharedse/impl/http/MountedContainerResponder.java
package com.ustadmobile.port.sharedse.impl.http;

import com.ustadmobile.lib.db.entities.Container;
import com.ustadmobile.lib.db.entities.ContainerEntryWithContainerEntryFile;
import com.ustadmobile.port.sharedse.container.ContainerManager;

import org.xmlpull.v1.XmlPullParserException;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Map;
import java.util.regex.Pattern;

import fi.iki.elonen.NanoHTTPD;
import fi.iki.elonen.router.RouterNanoHTTPD;

public class MountedContainerResponder extends FileResponder implements RouterNanoHTTPD.UriResponder {

    /**
     * The string that is added
     */
    public static final String URI_ROUTE_POSTFIX = "(.)+";

    private static final ArrayList<String> HTML_EXTENSIONS = new ArrayList<>();

    static {
        HTML_EXTENSIONS.add("xhtml");
        HTML_EXTENSIONS.add("html");
        HTML_EXTENSIONS.add("htm");
    }

    public static class MountedZipFilter {

        public MountedZipFilter(Pattern pattern, String replacement) {
            this.pattern = pattern;
            this.replacement = replacement;
        }

        public Pattern pattern;

        public String replacement;
    }

    public static class FilteredHtmlSource implements FileResponder.IFileSource {

        private FileResponder.IFileSource src;

        private ByteArrayInputStream inputStream;

        private long length = -1;

        private String scriptPath;

        public FilteredHtmlSource(FileResponder.IFileSource src, String scriptPath) {
            this.src = src;
            this.scriptPath = scriptPath;
        }

        @Override
        public long getLength() {
            try { getInputStream(); }
            catch(IOException e) {}
            return length;
        }

        @Override
        public long getLastModifiedTime() {
            return src.getLastModifiedTime();
        }

        @Override
        public InputStream getInputStream() throws IOException {
            if(inputStream == null) {
                //init and filter
                InputStream srcIn = src.getInputStream();
                try {
                    EpubHtmlFilterSerializer filterSerializer = new EpubHtmlFilterSerializer();
                    filterSerializer.setScriptSrcToAdd(scriptPath);
                    filterSerializer.setIntput(srcIn);
                    byte[] filteredBytes = filterSerializer.getOutput();
                    length = filteredBytes.length;
                    inputStream = new ByteArrayInputStream(filteredBytes);
                }catch(XmlPullParserException x) {
                    throw new IOException(x);
                }
            }
            return inputStream;
        }

        @Override
        public boolean exists() {
            return src.exists();
        }

        @Override
        public String getName() {
            return src.getName();
        }
    }


    @Override
    public NanoHTTPD.Response get(RouterNanoHTTPD.UriResource uriResource, Map<String, String> urlParams, NanoHTTPD.IHTTPSession session) {
        String requestUri = RouterNanoHTTPD.normalizeUri(session.getUri());
        String pathInContainer = requestUri.substring(
                uriResource.getUri().length() - URI_ROUTE_POSTFIX.length());
        ContainerManager container = uriResource.initParameter(0, ContainerManager.class);
        ContainerEntryWithContainerEntryFile entry = container.getEntry(pathInContainer);
        if(entry == null){
            return NanoHTTPD.newFixedLengthResponse(NanoHTTPD.Response.Status.NOT_FOUND,
                    "text/plain", "Entry not found in container");
        }

        return newResponseFromFile(uriResource, session,
                new FileSource(new File(entry.getContainerEntryFile().getCefPath())));
    }

    @Override
    public NanoHTTPD.Response put(RouterNanoHTTPD.UriResource uriResource, Map<String, String> urlParams, NanoHTTPD.IHTTPSession session) {
        return null;
    }

    @Override
    public NanoHTTPD.Response post(RouterNanoHTTPD.UriResource uriResource, Map<String, String> urlParams, NanoHTTPD.IHTTPSession session) {
        return null;
    }

    @Override
    public NanoHTTPD.Response delete(RouterNanoHTTPD.UriResource uriResource, Map<String, String> urlParams, NanoHTTPD.IHTTPSession session) {
        return null;
    }

    @Override
    public NanoHTTPD.Response other(String method, RouterNanoHTTPD.UriResource uriResource, Map<String, String> urlParams, NanoHTTPD.IHTTPSession session) {
        return null;
    }
}

File: sharedse/src/main/java/com/ustadmobile/port/sharedse/impl/http/MountedContainerResponder.kt
package com.ustadmobile.port.sharedse.impl.http

import com.google.gson.reflect.TypeToken
import com.ustadmobile.lib.db.entities.Container
import com.ustadmobile.lib.db.entities.ContainerEntryWithContainerEntryFile
import com.ustadmobile.port.sharedse.container.ContainerManager

import org.xmlpull.v1.XmlPullParserException

import java.io.ByteArrayInputStream
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.net.URI
import java.net.URISyntaxException
import java.util.ArrayList
import java.util.regex.Pattern

import fi.iki.elonen.NanoHTTPD
import fi.iki.elonen.router.RouterNanoHTTPD

class MountedContainerResponder : FileResponder(), RouterNanoHTTPD.UriResponder {

    interface MountedContainerFilter {

        fun filterResponse(responseIn: NanoHTTPD.Response,
                           uriResource: RouterNanoHTTPD.UriResource,
                           urlParams: Map<String, String>,
                           session: NanoHTTPD.IHTTPSession): NanoHTTPD.Response

    }

    @Deprecated("")
    class MountedZipFilter(var pattern: Pattern, var replacement: String)

    class FilteredHtmlSource(private val src: FileResponder.IFileSource, private val scriptPath: String) : FileResponder.IFileSource {

        private var inputStream: ByteArrayInputStream? = null

        private var length: Long = -1

        override val lastModifiedTime: Long
            get() = src.lastModifiedTime

        override val name: String
            get() = src.name

        override fun getLength(): Long {
            try {
                getInputStream()
            } catch (e: IOException) {
            }

            return length
        }

        @Throws(IOException::class)
        override fun getInputStream(): InputStream {
            if (inputStream == null) {
                //init and filter
                val srcIn = src.inputStream
                try {
                    val filterSerializer = EpubHtmlFilterSerializer()
                    filterSerializer.scriptSrcToAdd = scriptPath
                    filterSerializer.setIntput(srcIn)
                    val filteredBytes = filterSerializer.output
                    length = filteredBytes.size.toLong()
                    inputStream = ByteArrayInputStream(filteredBytes)
                } catch (x: XmlPullParserException) {
                    throw IOException(x)
                }

            }
            return inputStream
        }

        override fun exists(): Boolean {
            return src.exists()
        }
    }


    override fun get(uriResource: RouterNanoHTTPD.UriResource, urlParams: Map<String, String>, session: NanoHTTPD.IHTTPSession): NanoHTTPD.Response {
        try {
            var requestUri = RouterNanoHTTPD.normalizeUri(session.uri)
            requestUri = URI(requestUri).normalize().toString()
            val pathInContainer = requestUri.substring(
                    uriResource.uri.length - URI_ROUTE_POSTFIX.length)
            val container = uriResource.initParameter(0, ContainerManager::class.java)
            val entry = container.getEntry(pathInContainer)
                    ?: return NanoHTTPD.newFixedLengthResponse(NanoHTTPD.Response.Status.NOT_FOUND,
                            "text/plain", "Entry not found in container")

            val filterList = uriResource.initParameter(1, List<*>::class.java)
            var response = FileResponder.newResponseFromFile(uriResource, session,
                    FileResponder.FileSource(File(entry.containerEntryFile!!.cefPath!!)))
            for (filter in filterList) {
                response = filter.filterResponse(response, uriResource, urlParams, session)
            }

            return response
        } catch (e: URISyntaxException) {
            e.printStackTrace()
            return NanoHTTPD.newFixedLengthResponse(NanoHTTPD.Response.Status.BAD_REQUEST,
                    "text/plain", "URISyntax error: $e")
        }

    }

    override fun put(uriResource: RouterNanoHTTPD.UriResource, urlParams: Map<String, String>, session: NanoHTTPD.IHTTPSession): NanoHTTPD.Response? {
        return null
    }

    override fun post(uriResource: RouterNanoHTTPD.UriResource, urlParams: Map<String, String>, session: NanoHTTPD.IHTTPSession): NanoHTTPD.Response? {
        return null
    }

    override fun delete(uriResource: RouterNanoHTTPD.UriResource, urlParams: Map<String, String>, session: NanoHTTPD.IHTTPSession): NanoHTTPD.Response? {
        return null
    }

    override fun other(method: String, uriResource: RouterNanoHTTPD.UriResource, urlParams: Map<String, String>, session: NanoHTTPD.IHTTPSession): NanoHTTPD.Response? {
        return null
    }

    companion object {

        /**
         * The string that is added
         */
        val URI_ROUTE_POSTFIX = "(.)+"

        private val HTML_EXTENSIONS = ArrayList<String>()

        init {
            HTML_EXTENSIONS.add("xhtml")
            HTML_EXTENSIONS.add("html")
            HTML_EXTENSIONS.add("htm")
        }
    }
}
